### ISSUE TITLE
(OIHM Modelling): Sale of tradeable goods not executed as expected

### ISSUE THREAD

**Issue by daiClegg** [dbrownukk/EFD_v2#529]
Instance:  QA
App: OIHM/Modelling
Module: Livelihoods Modelling
Scenario: Katakwi - OIHM
Study: Katakwi - NIMFRU - Anyangabella
HH: 91

1. Run model execution with coping strategy for the above Scenario & HH
2. Open the data extract.
> note the Rule 'Sell Bicycle' results in no increase in income.

![image](https://user-images.githubusercontent.com/30879677/98038750-905b9680-1e15-11eb-8052-117232ec7613.png)

3. In HH module, view detail of Katakwi - NIMFRU - Anyangabella, HH 91, navigate to Assets/Tradeable Goods
> note the HH owns a bicycle and there is a non-zero value for the bicycle
4. In Study module, view detail view of Katakwi - NIMFRU - Anyangabella, navigate to the Coping Strategy
> note Sell Bicycle has no constraints on the rule.

Given the above, i had expected the HH to sell its bicycle.

I suspect the spec was ambiguous (or wrong) on the subject of sale of tradeable goods, but i'm not sure what else could trigger the sale.

**Comment by dbrownukk** [comment]
Review, document calculation and correct

**Comment by dbrownukk** [comment]
A possible solution to use Assets within Coping Strategy

Extend Expandability Rule to support the use of  Tradeable Goods, Cash & Food Stores as a Coping Strategy

For these types we have 

1. Maximum %age that can be disposed
2. What %age of the disposal is consumed and what is sold
3. Price sold at   - not needed for Cash..

Then include in Modelling report, Change Scenario + Coping Strategy and show as per existing Coping Strategy sold items.

?

**Comment by celiapetty** [comment]
Interesting-that sounds about right.

**Comment by celiapetty** [comment]
Reviewing discussion so far: the proposed approach makes sense. Rules can be explained in training ie
1. Food stores-up to 100% can be used as they are stored against future need.
2. Cash-not often collected but and as John points out earlier mostly held by better off.  For clarification: will user be able to set the ceiling or does this need to be pre-programmed . Ideally user defines ceiling for cash (cf livestock- if you loose all your capital this might make your future business non viable. Thinking more of future urban IHM studies here.
3. Land-ideally user defines ceiling based on local circumstances/type of farming. If not possible, suggest max sale of land (say) 60%
4. Trees. As with land: if wealth is in forestry it would make sense to put in a ceiling of (say) 60%
5. Tradeable goods. Most problematic as teams are trained to list 'productive' assets here (eg ploughs/other agricultural tools, tools of trade eg carpentry, sewing machine. Not tables, chairs etc) If items like motor bikes are sold, this is likely to be a distress sale resulting in loss of income from motor bike taxi. Also -see John's comments above- We could set a very low ceiling?

**Comment by DrJohnSeaman** [comment]
Food stores-up to 100% can be used as they are stored against future need. OK

Trees.  if wealth is in forestry it would make sense to put in a ceiling of (say) 60%. Not sure about this - there are situations where people have say a few trees and would use 100% if they had to e.g. coppiced eucalyptus in Ethiopia which are ept as a fall back. They grow back to saleable poles pretty quickly.

Tradeable goods. Most problematic as teams are trained to list 'productive' assets here (eg ploughs/other agricultural tools, tools of trade eg carpentry, sewing machine. Not tables, chairs etc) If items like motor bikes are sold, this is likely to be a distress sale resulting in loss of income from motor bike taxi. A very low ceiling.

Cash-I think the user should set the ceiling - its use is lilkey to be a hypothetical her than a real operational modelling.
Land-ideally user defines ceiling based on local circumstances/type of farming. If not possible, suggest max sale of land (say) 60% OK

**Comment by daiClegg** [comment]
Thanks for the detailed input @DrJohnSeaman 

The wider modelling extensions are being tracked in Issue #443 where i am building a set of rules for Change Scenario Elements and Expansion Rules.

**Comment by dbrownukk** [comment]
Assets - Cash, Food Stores and Tradeable Goods now added to Modelling Report - Cash and Food Items

I know wider extension is proposed but this enhancement is just about there.

A Study / Community Expandability Rule now has 

Disposal Maximum Percent
Disposal Price
Disposal Percent consumed
Disposal Percent Sold

Added to Modelling report as Coping STrategy as before

Should there be a type of Expandability rule? Could a Coping Strategy have the above and 

Expandability Increase Limit, Expandability Limit

with both taken into account in modelling report coping strategy run. 

To test an Asset + Expandability Rule for Asset is required.

deployed to IDAPS_Livelihoods_TESTING

**Comment by daiClegg** [comment]
Blocked on #638 cos i need to check out WGI detail to create coping strategies that will affect WGs appropriately to test this enhancement

**Comment by daiClegg** [comment]
unblocked. need to find a WG on the edge, then sewt up Community coping strategy to include
one each of the new categories supported.
note will need to manipulate the Assets & Resources of my target WG to make sure the coping strategy will help them
(ie need a tree, some cash, a food store etc.

**Comment by daiClegg** [comment]
Model: Modelling Extensions Test
Village: Ku
WG: Poor

This WG has -ive DI (manipulated) so the ChangeScenario - though it doesn't affect them much is enough to trigger the CopingStrategy.

TBD: 
1. set up Community Coping Strategy to include new rule types, 
2. set up WG assets & resources to test

**Comment by daiClegg** [comment]
> Added the Expandability Rules to include:
1.  Spend Cash
2.  Increase hours on Agricultural Labour
3. Sell bicycle
> Edited the Ku Poor WGI to add these reosurces

DI was negative after Change Scenario, but no Expandability Rules were applied. 

**Next step:**
Review whether i set appropriate values in the Expandability Rules.

**Comment by daiClegg** [comment]
not sure why i thought Employment was included here - that was a red herring that i set for myself.

**Comment by daiClegg** [comment]
Instance: Testing
Module: Livelihoods Modelling
Model: Modelling Extensions Test
Project: Nimfru
LZ: Katakwi - Fishing & Livestock Zone
Village: Ku
WG: MIddle

1. In the Community module note Expandability rules for 'Ku' include 'Spend Cash' & 'Sell Bicycle'
> Note properties set to not impose any constraints ( i think)
2. In WGI Module note 'Ku' MIddle & Poor both have Cash & a Bicycle recorded as Assets 
3. In Livelihoods model, run Change Scenario & Coping Strategy for model 'Modelling Extensions Test' 
4. Select Village 'Ku'
5. In downloaded data extract view '425 WG DI with Coping Strategy' tab
> Note the impact of the expandability rules are not as the rules and assets would seem to predict
> ( no impact on 'Poor' WG, odd impacts on 'Middle' WG)

Annotated data extract uploaded here:
[KuVillage4Rules.xlsx](https://github.com/dbrownukk/EFD_v2/files/6504186/KuVillage4Rules.xlsx)
Data extract from a previous run (with an extraneous Expandability rule) uploaded here
[KuVillage5Rules.xlsx](https://github.com/dbrownukk/EFD_v2/files/6504191/KuVillage5Rules.xlsx)
This is included because annotations in the other upload reference some values in this one

**Comment by dbrownukk** [comment]
Ah. Misinterpretation. I have used Disposal Percent Consumed for FI and Disposal Percent Sold for CI

Disposal Max Percent not used as yet. 

Before I change anything how does Percent Consumed / Sold relate to Max Percent? 

Would it be dispose of 100% and of that x% is consumed and x% sold. At the moment in the above example 0 % is consumed or sold.

**Comment by daiClegg** [comment]
I realise i haven't added sale of assets (including tradable goods) to the spreadsheet of expandability rules attached to #443.
That's an omission that i will correct. 
And will extract the rules relevant for assets here 

**TBD**

**Comment by daiClegg** [comment]
From the additional fields above, are % consumed & % sold both included to allow for a % other use?

**Comment by daiClegg** [comment]
revised rules attached to #443.

to be discussed...

**Comment by dbrownukk** [comment]
A specific example:-  A Household has a Tradeable asset of a Bicycle value 25,000

The Study has an Expandability Rule of Other Tradeable Goods Bicycle, Expandability Limit 100%, Disposal Price 30,000

The Modelling report run will report the sale of 1 bicycle at 30,000 increasing cash income by 30000

Ready for testing and deployed in  IDAPS_Livelihoods_TESTING

**Comment by daiClegg** [comment]
I added back spend savings and sell bicycle as extendability rules for Community Ku in the test scenario described above (18 may).

Using the same run sequence, If i run coping strategy just for this village i get an error message
 > Impossible to execute Run action: null
If i run coping strategy with Ku and any other combination of other villages (i tried several), it runs ok
**BUT** the two additional expandability rules for Ku are ignored.

I realise we have not discussed a coping strategy run for villlages with different expandability rules,
Right now i'm not sure what the answer is. Needs more thought, but there is an error anyway.

**Comment by daiClegg** [comment]
Can we discuss please @dbrownukk

**Comment by dbrownukk** [comment]
Pulling data from idaps server to local dev server to test..

**Comment by dbrownukk** [comment]
Village KU Poor has a Cash Asset. The currency check in modelling report was checking 'currency entered' not the Cash RST currency drop down. In Poor this was blank.

Fixed, deployed to IDAPS_Livelihoods_TESTING

**Comment by daiClegg** [comment]
Re did tests from y'day. **Tested OK**
Ie ran successfully with no errors.
However...

If i run just Ku all 4 expandability rules are applied and reported in the data extract.
But if i report all the villages then only the rule the others have in place are applied.

That's what i am now concerned about. What do i expect should happen? Is that right?
Hence, to be discussed

**Comment by daiClegg** [comment]
Further Testing (scenario as above)
- checked the full LZ report to find what is the first village reported on: it's Ae
- added a 3rd expandability rule to Ae
-  ran coping strategy just for Ae: ran as expected.
-  ran coping strategy for the whole LZ: only 2 rules are applied to Ae & Ku, both of which have further rules

I added the rule to Ae to test an idea that the set of rules to be applied was being picked up from the first village in the query. It may be but if it is then it's not that one.

I checked through all the villages in that LZ and most have 2 rules (Sell Goats, Sell Groundnuts) and those are the rules applied. even to villages (like Ae that does not have the Sell Groundnuts rule). I don't know why those rules are the chosen ones.

We should not apply an expandability rule to a village that does not have that rule defined.
We should apply all expandability rules that a village does have defined.

I think our options are: 

1. A report run is constrained to just run a set of villages with the same rules.
2. A report run applies the union set of all rules, but only applies each one to the villages that define it.

In the first case the report would have to do that check pre-run. But it would be very difficult for the user to work out which rules apply to which villages and so with ones to include unless the UI helps. for example by listing the RSTs in its coping strategy as a concatenated string to the right of the report popup village list, something like this:
![image](https://user-images.githubusercontent.com/30879677/125163079-46d10180-e183-11eb-8bee-8b0134dd62ba.png)

In the second case we would need to include the coping strategies in thein the data extract, so the user can discern which villages are doing what. Currently the coping strategy is show=n as one universal set of rules in the Summary Tab.
It could equally be shown for each village in columns E, F, G etc in the same rows as the village names.

I will ask Celia which approach is more desirable.

**Comment by daiClegg** [comment]
I just realised that if we have the UI enhancement above for option 1 and implementation of option 2, the user can have it both ways.  I may have jusdt convinced myself that the most expensive solution is the best.

**Comment by dbrownukk** [comment]
I can see how this happened. I think I did IHM first and implemented HEA coping strategy in the same way. Big mistake. 

Before I fix this, definitely need to know what scenario is best. Does a report with different Expandability Rules per each Community and thus Village make sense? Is a report that says that one village can sell bicycles and another has a drop in price of groundnuts, produce data that is of any use? 

Perhaps better on running an HEA coping strategy report to select LZ as now and then select which Expandability rules to apply to all villages to make a consistent report. 

@dai for discussion.

**Comment by daiClegg** [comment]
Meeting with Celia, John & James arranged.
Notes from pre-meeting discussion:

Celia:
It sounds as if there's a consensus that it's useful to allow reporting of coping strategies at community level for more detailed studies, but that for large area food security assessments these should be agreed at LZ level. 
In terms of the software (Dai, correct me if I'm wrong),  this means that for both large area/rapid and more focussed studies (cf HyCristal and Nimfru) users will have to enter coping strategies for each village site. They will be entering the same coping strategies at each village site for the large area studies, but for detailed studies, it's possible that different coping strategies will be registered. Dai-what are the implications of this for reporting? 
Will we-in effect- be creating sub LZs? 
Possibly report on the whole LZ for DI/main income sources etc and sub-LZs for coping?

**Comment by daiClegg** [comment]
My responses to he questions asked of me:

1. entering coping strategies for each village: 
> yes, currently you have to do that. But in #479 i suggest a way to minimise the data entry but still get the flexibility.
2. reporting sub-sets of the communities.
> currently you can report any arbitrary sub-set of the villages in a LZ, just by selecting or not selecting them in the modelling reports dialog.
> but you would have to have gone through the Community records for each Village and made a note beforehand which villages were to be included
> however in the proposal i mocked up in this thread (10 days before this entry) the expandability rules are shown so the user could easily see which villages they want to include in any model run.
> that proposal would support the requirement that the user can do:
-  drill-down (selected Village or Villages) based in which oners have te same Coping Strategy
-  LZ-level reporting

**Comment by daiClegg** [comment]
Notes from a call with Celia, Jame & John (21/jul/21). This is a very brief summary focussing on the impact on the OHEA software.

- The core use case is using coping strategies to understand what might be expected to be the outcome  of a shock (outcome analysis). This is typically part of a large scale food security study. IN this case the unit of analysis id not going to be smaller than the LZ. 

- There are situations where it is valuable to drill down into what might be called 'sub-zones' to understand how different responses (coping strategies) might be equally, or more or less effective.


So my supposition, described in previous posts, is that:
1.  copy Coping Strategy would be useful because it allows us to maintain variation but lower the key-stroke load on the user when they set it up.
2.displaying the coping strategy for each village in the modelling report dialogue would have value because it would allow the user to see and use the "sub zones" of similarly responsive villages.
3. we still have to decide how the modelling and visualisation would deal with applying only the appropriate rules for a village.

In the light of 3, we need an engineering discussion.

**Comment by daiClegg** [comment]
Note an additional visualisation of a matrix of villages & rules could show (tick in the intersection cell) which rules apply where.

**Comment by daiClegg** [comment]
Closing this issue, since #652 is about the implementation of the new feature that deals with the perceived anomaly reported here.
